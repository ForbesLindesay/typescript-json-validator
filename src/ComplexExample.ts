export enum MyEnum {
  ValueA,
  ValueB,
  ValueC,
}

export interface TypeA {
  id: number;
  value: string;
}

export interface TypeB {
  id: number | null;
  /**
   * @format date-time
   */
  value: string | null;
}

export interface RequestA {
  query: TypeA;
  body: TypeB;
  params: {e: MyEnum};
}

export interface RequestB {
  query: TypeA;
}
