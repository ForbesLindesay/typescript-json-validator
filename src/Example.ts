export default interface ExampleType {
  value: string;
  /**
   * @TJS-format email
   */
  email?: string;
  /**
   * @default 42
   */
  answer: number;
}
