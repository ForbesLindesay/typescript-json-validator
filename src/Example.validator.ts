/* tslint:disable */
// generated by typescript-json-validator
import {inspect} from 'util';
import Ajv = require('ajv');
import ExampleType from './Example';
export const ajv = new Ajv({
  allErrors: true,
  coerceTypes: false,
  format: 'fast',
  nullable: true,
  unicode: true,
  uniqueItems: true,
  useDefaults: true,
});

ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));

export {ExampleType};
export const ExampleTypeSchema = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  defaultProperties: [],
  properties: {
    answer: {
      default: 42,
      type: 'number',
    },
    email: {
      format: 'email',
      type: 'string',
    },
    value: {
      type: 'string',
    },
  },
  required: ['answer', 'value'],
  type: 'object',
};
export type ValidateFunction<T> = ((data: unknown) => data is T) &
  Pick<Ajv.ValidateFunction, 'errors'>;
const rawValidateExampleType = ajv.compile(
  ExampleTypeSchema,
) as ValidateFunction<ExampleType>;
export default function validate(value: unknown): ExampleType {
  if (rawValidateExampleType(value)) {
    return value;
  } else {
    throw new Error(
      ajv.errorsText(rawValidateExampleType.errors, {dataVar: 'ExampleType'}) +
        '\n\n' +
        inspect(value),
    );
  }
}
