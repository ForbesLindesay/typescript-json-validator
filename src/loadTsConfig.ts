import loadTsconfig from 'tsconfig-loader';

export default function loadTsConfig(cwd: string = process.cwd()) {
  const result = loadTsconfig({cwd});
  const compilerOptions = result?.tsConfig.compilerOptions || {};
  if (
    compilerOptions.experimentalDecorators === false &&
    compilerOptions.emitDecoratorMetadata === undefined
  ) {
    // typescript-json-schema sets emitDecoratorMetadata by default
    // we need to disable it if experimentalDecorators support is off
    compilerOptions.emitDecoratorMetadata = false;
  }
  if (compilerOptions.composite) {
    // the composite setting adds a few constraints that cause us all manner of problems
    compilerOptions.composite = false;
  }
  compilerOptions.incremental = false;

  // since composite and incremental are false, Typescript will not accept tsBuildInfoFile
  // https://github.com/microsoft/TypeScript/blob/dcb763f62435ebb015e7fa405eb067de3254f217/src/compiler/program.ts#L2847
  delete compilerOptions.tsBuildInfoFile;

  return compilerOptions;
}
