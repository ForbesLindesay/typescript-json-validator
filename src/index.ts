import {writeFileSync} from 'fs';
import {basename} from 'path';
import {parseArgs} from './parseArgs';
import parse from './parse';
import {
  printSingleTypeValidator,
  printTypeCollectionValidator,
} from './printValidator';
import prettierFile from './prettierFile';
import loadTsConfig from './loadTsConfig';
import normalizeSchema from './normalizeSchema';

export {
  parse,
  parseArgs,
  printSingleTypeValidator,
  printTypeCollectionValidator,
};

export default function run(args?: string[]) {
  const {files, options} = parseArgs(args);
  const tsConfig = loadTsConfig();
  const parsed = parse(
    files.map(f => f.fileName),
    tsConfig,
    options.schema,
  );

  files.forEach(({fileName, typeName}) => {
    const outputFileName = fileName.replace(/\.tsx?$/, '.validator.ts');
    let validator: string;
    if (typeName) {
      const schema = parsed.getType(typeName);
      validator = printSingleTypeValidator(
        typeName,
        options.useNamedExport,
        normalizeSchema(schema),
        `./${basename(fileName, /\.ts$/.test(fileName) ? '.ts' : '.tsx')}`,
        tsConfig,
        options.ajv,
      );
    } else {
      const {symbols, schema} = parsed.getAllTypes();
      validator = printTypeCollectionValidator(
        symbols,
        normalizeSchema(schema),
        `./${basename(fileName, /\.ts$/.test(fileName) ? '.ts' : '.tsx')}`,
        tsConfig,
        options.ajv,
      );
    }
    writeFileSync(outputFileName, validator);
    prettierFile(outputFileName);
  });
}
