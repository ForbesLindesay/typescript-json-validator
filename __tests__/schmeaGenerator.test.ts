import {createParser as parse} from '../src/schemaGenerator';
import Ajv from 'ajv';
import loadTsConfig from '../src/loadTsConfig';
import * as path from 'path';
import {isBoolean} from 'lodash';

const complexPath = path.join(
  __dirname,
  'build-parameters/src/ComplexExample.ts',
);

type ParseTestContext = {
  file: string;
  type: string;
};

type ValidateTestContext = {
  file: string;
  types: ValidateSchemaContext[];
};
type ValidateSchemaContext = {
  schemaRef: string;
  validations: {value: any; toBe: boolean}[];
};

function getPathToFile(file: string): string {
  return path.join(__dirname, 'build-parameters', 'src', file);
}

test.each<ParseTestContext>([
  {file: 'ComplexExample.ts', type: 'TypeA'},
  {file: 'Example.ts', type: 'ExampleClass'},
  {file: 'DisjointUnionExample.ts', type: 'EntityOne'},
  {file: 'OmitExample.ts', type: 'ISchema'},
])('Use schemaGenerator to parse $file', (tc: ParseTestContext) => {
  const filePath = getPathToFile(tc.file);
  const parser = parse(filePath, loadTsConfig());
  expect(parser.getAllTypes()).toMatchSnapshot();
  expect(parser.getType(tc.type)).toMatchSnapshot();
  expect(parser.getTypeExport(tc.type)).toMatchSnapshot();
});

test.each<ValidateTestContext>([
  {
    file: 'OmitExample.ts',
    types: [
      {
        schemaRef: 'root#/definitions/ISchema',
        validations: [
          {value: {id: '1', schema: 1}, toBe: false},
          {value: {id: '1', schema: {type: 'string'}}, toBe: true},
        ],
      },
    ],
  },
  {
    file: 'ComplexExample.ts',
    types: [
      {
        schemaRef: 'root#/definitions/MyEnum',
        validations: [
          {value: 1, toBe: true},
          {value: 10, toBe: false},
        ],
      },
      {
        schemaRef: 'root#/definitions/RequestA',
        validations: [
          {value: {query: {id: 'x', value: 'y'}, params: {e: 42}}, toBe: false},
        ],
      },
    ],
  },

  {
    file: 'Example.ts',
    types: [
      {
        schemaRef: 'root#/definitions/ExampleClass',
        validations: [
          {value: {value: 'test'}, toBe: true},
          {value: {value: 'test', email: 'robert@g.com'}, toBe: true},
          {
            value: {value: 'test', email: 'robert@g.com', answer: 50},
            toBe: true,
          },
          {value: {value: 'test', email: 'robert', answer: 50}, toBe: false},
          {
            value: {value: 'test', email: 'robert@g.com', answer: '50'},
            toBe: false,
          },
        ],
      },
      {
        schemaRef: 'root#/definitions/ExampleType',
        validations: [
          {value: {value: 'test'}, toBe: true},
          {value: {value: 'test', email: 'robert@g.com'}, toBe: true},
          {
            value: {value: 'test', email: 'robert@g.com', answer: 50},
            toBe: true,
          },
          {value: {value: 'test', email: 'robert', answer: 50}, toBe: false},
          {
            value: {value: 'test', email: 'robert@g.com', answer: '50'},
            toBe: false,
          },
        ],
      },
    ],
  },
])('Validating with ajv for $file', (tc: ValidateTestContext) => {
  const filePath = getPathToFile(tc.file);
  const parsed = parse(filePath, loadTsConfig());
  const {schema} = parsed.getAllTypes();
  const ajv = new Ajv({coerceTypes: false, allErrors: true, useDefaults: true});
  ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));
  if (isBoolean(schema)) {
    fail();
  }
  ajv.addSchema(schema, 'root');

  tc.types.forEach((type) => {
    const validator = ajv.getSchema(type.schemaRef)!;
    if (!validator) {
      console.log(`No validator for ${type.schemaRef}`);
    }
    expect(validator).toBeDefined();
    type.validations.forEach((v) => {
      if (validator(v.value) !== v.toBe) {
        console.log(
          `Failed validation ${JSON.stringify(v.value)}, expected: ${v.toBe}`,
        );
      }
      expect(validator(v.value)).toBe(v.toBe);
      if (!v.toBe) {
        expect(
          ajv.errorsText(validator.errors, {dataVar: 'x'}),
        ).toMatchSnapshot();
      }
    });
  });
});

test('ajv', () => {
  const parsed = parse(complexPath, loadTsConfig());
  const {schema} = parsed.getAllTypes();
  const ajv = new Ajv({coerceTypes: false, allErrors: true, useDefaults: true});
  if (isBoolean(schema)) {
    fail();
  }
  ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));
  ajv.addSchema(schema, 'root');

  const validateMyEnum = ajv.getSchema('root#/definitions/MyEnum')!;
  expect(validateMyEnum(1)).toBe(true);
  expect(validateMyEnum(10)).toBe(false);
  expect(
    ajv.errorsText(validateMyEnum.errors, {dataVar: 'x'}),
  ).toMatchInlineSnapshot(`"x should be equal to one of the allowed values"`);

  const validateRequestA = ajv.getSchema('root#/definitions/RequestA')!;
  expect(
    validateRequestA({query: {id: 'x', value: 'y'}, params: {e: 42}}),
  ).toBe(false);
  expect(
    ajv.errorsText(validateRequestA.errors, {dataVar: 'req'}),
  ).toMatchInlineSnapshot(
    `"req.query.id should be number, req should have required property 'body', req.params.e should be equal to one of the allowed values"`,
  );
});
