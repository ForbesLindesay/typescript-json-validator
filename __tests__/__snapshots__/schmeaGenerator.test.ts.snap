// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use schemaGenerator to parse ComplexExample.ts 1`] = `
Object {
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {
      "MyEnum": Object {
        "enum": Array [
          0,
          1,
          2,
        ],
        "type": "number",
      },
      "RequestA": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "body": Object {
            "$ref": "#/definitions/TypeB",
          },
          "params": Object {
            "additionalProperties": undefined,
            "properties": Object {
              "e": Object {
                "$ref": "#/definitions/MyEnum",
              },
            },
            "required": Array [
              "e",
            ],
            "type": "object",
          },
          "query": Object {
            "$ref": "#/definitions/TypeA",
          },
        },
        "required": Array [
          "query",
          "body",
          "params",
        ],
        "type": "object",
      },
      "RequestB": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "query": Object {
            "$ref": "#/definitions/TypeA",
          },
        },
        "required": Array [
          "query",
        ],
        "type": "object",
      },
      "TypeA": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "id": Object {
            "type": "number",
          },
          "value": Object {
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "value",
        ],
        "type": "object",
      },
      "TypeB": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "id": Object {
            "type": Array [
              "number",
              "null",
            ],
          },
          "value": Object {
            "format": "date-time",
            "type": Array [
              "string",
              "null",
            ],
          },
        },
        "required": Array [
          "id",
          "value",
        ],
        "type": "object",
      },
    },
  },
  "symbols": Array [
    "MyEnum",
    "TypeA",
    "TypeB",
    "RequestA",
    "RequestB",
  ],
}
`;

exports[`Use schemaGenerator to parse ComplexExample.ts 2`] = `
Object {
  "$ref": "#/definitions/TypeA",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "TypeA": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "id": Object {
          "type": "number",
        },
        "value": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "value",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Use schemaGenerator to parse ComplexExample.ts 3`] = `"TypeA"`;

exports[`Use schemaGenerator to parse DisjointUnionExample.ts 1`] = `
Object {
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {
      "Entity": Object {
        "anyOf": Array [
          Object {
            "$ref": "#/definitions/EntityOne",
          },
          Object {
            "$ref": "#/definitions/EntityTwo",
          },
          Object {
            "additionalProperties": undefined,
            "properties": Object {
              "baz": Object {
                "type": "number",
              },
              "type": Object {
                "const": "TypeThree",
                "type": "string",
              },
            },
            "required": Array [
              "type",
              "baz",
            ],
            "type": "object",
          },
        ],
      },
      "EntityOne": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "foo": Object {
            "type": "string",
          },
          "type": Object {
            "const": "TypeOne",
            "type": "string",
          },
        },
        "required": Array [
          "type",
          "foo",
        ],
        "type": "object",
      },
      "EntityTwo": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "bar": Object {
            "type": "string",
          },
          "type": Object {
            "const": "TypeTwo",
            "type": "string",
          },
        },
        "required": Array [
          "type",
          "bar",
        ],
        "type": "object",
      },
      "EntityTypes": Object {
        "enum": Array [
          "TypeOne",
          "TypeTwo",
          "TypeThree",
        ],
        "type": "string",
      },
      "Value": Object {
        "anyOf": Array [
          Object {
            "additionalProperties": undefined,
            "properties": Object {
              "foo": Object {
                "type": "string",
              },
              "number": Object {
                "const": 0,
                "type": "number",
              },
            },
            "required": Array [
              "number",
              "foo",
            ],
            "type": "object",
          },
          Object {
            "additionalProperties": undefined,
            "properties": Object {
              "bar": Object {
                "type": "string",
              },
              "number": Object {
                "const": 1,
                "type": "number",
              },
            },
            "required": Array [
              "number",
              "bar",
            ],
            "type": "object",
          },
          Object {
            "additionalProperties": undefined,
            "properties": Object {
              "baz": Object {
                "type": "string",
              },
              "number": Object {
                "const": 2,
                "type": "number",
              },
            },
            "required": Array [
              "number",
              "baz",
            ],
            "type": "object",
          },
        ],
      },
    },
  },
  "symbols": Array [
    "EntityTypes",
    "EntityOne",
    "EntityTwo",
    "Entity",
    "Value",
  ],
}
`;

exports[`Use schemaGenerator to parse DisjointUnionExample.ts 2`] = `
Object {
  "$ref": "#/definitions/EntityOne",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "EntityOne": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "foo": Object {
          "type": "string",
        },
        "type": Object {
          "const": "TypeOne",
          "type": "string",
        },
      },
      "required": Array [
        "type",
        "foo",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Use schemaGenerator to parse DisjointUnionExample.ts 3`] = `"EntityOne"`;

exports[`Use schemaGenerator to parse Example.ts 1`] = `
Object {
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {
      "ExampleClass": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "answer": Object {
            "default": 42,
            "type": "number",
          },
          "email": Object {
            "format": "email",
            "type": "string",
          },
          "value": Object {
            "type": "string",
          },
        },
        "required": Array [
          "value",
          "answer",
        ],
        "type": "object",
      },
      "ExampleType": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "answer": Object {
            "default": 42,
            "type": "number",
          },
          "email": Object {
            "format": "email",
            "type": "string",
          },
          "value": Object {
            "type": "string",
          },
        },
        "required": Array [
          "value",
          "answer",
        ],
        "type": "object",
      },
      "Foo": Object {
        "type": "string",
      },
    },
  },
  "symbols": Array [
    "ExampleType",
    "ExampleClass",
    "Foo",
  ],
}
`;

exports[`Use schemaGenerator to parse Example.ts 2`] = `
Object {
  "$ref": "#/definitions/ExampleClass",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "ExampleClass": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "answer": Object {
          "default": 42,
          "type": "number",
        },
        "email": Object {
          "format": "email",
          "type": "string",
        },
        "value": Object {
          "type": "string",
        },
      },
      "required": Array [
        "value",
        "answer",
      ],
      "type": "object",
    },
  },
}
`;

exports[`Use schemaGenerator to parse Example.ts 3`] = `"ExampleClass"`;

exports[`Use schemaGenerator to parse OmitExample.ts 1`] = `
Object {
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {
      "ISchema": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "id": Object {
            "type": "string",
          },
          "schema": Object {
            "$ref": "#/definitions/JSONSchema7",
          },
          "uischema": Object {
            "$ref": "#/definitions/IUiSchemaNoReact",
          },
        },
        "required": Array [
          "id",
          "schema",
        ],
        "type": "object",
      },
      "IUiSchemaNoReact": Object {
        "additionalProperties": Object {},
        "properties": Object {
          "ui:field": Object {
            "type": "string",
          },
          "ui:options": Object {
            "additionalProperties": Object {
              "anyOf": Array [
                Object {
                  "type": "boolean",
                },
                Object {
                  "type": "number",
                },
                Object {
                  "type": "string",
                },
                Object {
                  "additionalProperties": Object {},
                  "type": "object",
                },
                Object {
                  "items": Object {},
                  "type": "array",
                },
                Object {
                  "type": "null",
                },
              ],
            },
            "type": "object",
          },
          "ui:order": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
          "ui:widget": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "JSONSchema7": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "$comment": Object {
            "type": "string",
          },
          "$defs": Object {
            "additionalProperties": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "object",
          },
          "$id": Object {
            "type": "string",
          },
          "$ref": Object {
            "type": "string",
          },
          "$schema": Object {
            "$ref": "#/definitions/JSONSchema7Version",
          },
          "additionalItems": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "additionalProperties": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "allOf": Object {
            "items": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "array",
          },
          "anyOf": Object {
            "items": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "array",
          },
          "const": Object {
            "$ref": "#/definitions/JSONSchema7Type",
          },
          "contains": Object {
            "$ref": "#/definitions/JSONSchema7",
          },
          "contentEncoding": Object {
            "type": "string",
          },
          "contentMediaType": Object {
            "type": "string",
          },
          "default": Object {
            "$ref": "#/definitions/JSONSchema7Type",
          },
          "definitions": Object {
            "additionalProperties": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "object",
          },
          "dependencies": Object {
            "additionalProperties": Object {
              "anyOf": Array [
                Object {
                  "$ref": "#/definitions/JSONSchema7Definition",
                },
                Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                },
              ],
            },
            "type": "object",
          },
          "description": Object {
            "type": "string",
          },
          "else": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "enum": Object {
            "items": Object {
              "$ref": "#/definitions/JSONSchema7Type",
            },
            "type": "array",
          },
          "examples": Object {
            "$ref": "#/definitions/JSONSchema7Type",
          },
          "exclusiveMaximum": Object {
            "type": "number",
          },
          "exclusiveMinimum": Object {
            "type": "number",
          },
          "format": Object {
            "type": "string",
          },
          "if": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "items": Object {
            "anyOf": Array [
              Object {
                "$ref": "#/definitions/JSONSchema7Definition",
              },
              Object {
                "items": Object {
                  "$ref": "#/definitions/JSONSchema7Definition",
                },
                "type": "array",
              },
            ],
          },
          "maxItems": Object {
            "type": "number",
          },
          "maxLength": Object {
            "type": "number",
          },
          "maxProperties": Object {
            "type": "number",
          },
          "maximum": Object {
            "type": "number",
          },
          "minItems": Object {
            "type": "number",
          },
          "minLength": Object {
            "type": "number",
          },
          "minProperties": Object {
            "type": "number",
          },
          "minimum": Object {
            "type": "number",
          },
          "multipleOf": Object {
            "type": "number",
          },
          "not": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "oneOf": Object {
            "items": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "array",
          },
          "pattern": Object {
            "type": "string",
          },
          "patternProperties": Object {
            "additionalProperties": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "object",
          },
          "properties": Object {
            "additionalProperties": Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            "type": "object",
          },
          "propertyNames": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "readOnly": Object {
            "type": "boolean",
          },
          "required": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
          },
          "then": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "title": Object {
            "type": "string",
          },
          "type": Object {
            "anyOf": Array [
              Object {
                "$ref": "#/definitions/JSONSchema7TypeName",
              },
              Object {
                "items": Object {
                  "$ref": "#/definitions/JSONSchema7TypeName",
                },
                "type": "array",
              },
            ],
          },
          "uniqueItems": Object {
            "type": "boolean",
          },
          "writeOnly": Object {
            "type": "boolean",
          },
        },
        "type": "object",
      },
      "JSONSchema7Array": Object {
        "additionalProperties": undefined,
        "properties": Object {
          "length": Object {
            "type": "number",
          },
        },
        "required": Array [
          "length",
        ],
        "type": "object",
      },
      "JSONSchema7Definition": Object {
        "anyOf": Array [
          Object {
            "$ref": "#/definitions/JSONSchema7",
          },
          Object {
            "type": "boolean",
          },
        ],
        "description": "JSON Schema v7",
      },
      "JSONSchema7Object": Object {
        "additionalProperties": Object {
          "$ref": "#/definitions/JSONSchema7Type",
        },
        "type": "object",
      },
      "JSONSchema7Type": Object {
        "anyOf": Array [
          Object {
            "type": "string",
          },
          Object {
            "type": "number",
          },
          Object {
            "type": "boolean",
          },
          Object {
            "$ref": "#/definitions/JSONSchema7Object",
          },
          Object {
            "$ref": "#/definitions/JSONSchema7Array",
          },
          Object {
            "type": "null",
          },
        ],
        "description": "Primitive type",
      },
      "JSONSchema7TypeName": Object {
        "description": "Primitive type",
        "enum": Array [
          "string",
          "number",
          "integer",
          "boolean",
          "object",
          "array",
          "null",
        ],
        "type": "string",
      },
      "JSONSchema7Version": Object {
        "description": "Meta schema

Recommended values:
- 'http://json-schema.org/schema#'
- 'http://json-schema.org/hyper-schema#'
- 'http://json-schema.org/draft-07/schema#'
- 'http://json-schema.org/draft-07/hyper-schema#'",
        "type": "string",
      },
    },
  },
  "symbols": Array [
    "IUiSchemaNoReact",
    "ISchema",
    "JSONSchema7",
    "JSONSchema7Version",
    "JSONSchema7Definition",
    "JSONSchema7TypeName",
    "JSONSchema7Type",
    "JSONSchema7Object",
    "JSONSchema7Array",
  ],
}
`;

exports[`Use schemaGenerator to parse OmitExample.ts 2`] = `
Object {
  "$ref": "#/definitions/ISchema",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "ISchema": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "schema": Object {
          "$ref": "#/definitions/JSONSchema7",
        },
        "uischema": Object {
          "$ref": "#/definitions/IUiSchemaNoReact",
        },
      },
      "required": Array [
        "id",
        "schema",
      ],
      "type": "object",
    },
    "IUiSchemaNoReact": Object {
      "additionalProperties": Object {},
      "properties": Object {
        "ui:field": Object {
          "type": "string",
        },
        "ui:options": Object {
          "additionalProperties": Object {
            "anyOf": Array [
              Object {
                "type": "boolean",
              },
              Object {
                "type": "number",
              },
              Object {
                "type": "string",
              },
              Object {
                "additionalProperties": Object {},
                "type": "object",
              },
              Object {
                "items": Object {},
                "type": "array",
              },
              Object {
                "type": "null",
              },
            ],
          },
          "type": "object",
        },
        "ui:order": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "ui:widget": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "JSONSchema7": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "$comment": Object {
          "type": "string",
        },
        "$defs": Object {
          "additionalProperties": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "object",
        },
        "$id": Object {
          "type": "string",
        },
        "$ref": Object {
          "type": "string",
        },
        "$schema": Object {
          "$ref": "#/definitions/JSONSchema7Version",
        },
        "additionalItems": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "additionalProperties": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "allOf": Object {
          "items": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "array",
        },
        "anyOf": Object {
          "items": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "array",
        },
        "const": Object {
          "$ref": "#/definitions/JSONSchema7Type",
        },
        "contains": Object {
          "$ref": "#/definitions/JSONSchema7",
        },
        "contentEncoding": Object {
          "type": "string",
        },
        "contentMediaType": Object {
          "type": "string",
        },
        "default": Object {
          "$ref": "#/definitions/JSONSchema7Type",
        },
        "definitions": Object {
          "additionalProperties": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "object",
        },
        "dependencies": Object {
          "additionalProperties": Object {
            "anyOf": Array [
              Object {
                "$ref": "#/definitions/JSONSchema7Definition",
              },
              Object {
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
            ],
          },
          "type": "object",
        },
        "description": Object {
          "type": "string",
        },
        "else": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "enum": Object {
          "items": Object {
            "$ref": "#/definitions/JSONSchema7Type",
          },
          "type": "array",
        },
        "examples": Object {
          "$ref": "#/definitions/JSONSchema7Type",
        },
        "exclusiveMaximum": Object {
          "type": "number",
        },
        "exclusiveMinimum": Object {
          "type": "number",
        },
        "format": Object {
          "type": "string",
        },
        "if": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "items": Object {
          "anyOf": Array [
            Object {
              "$ref": "#/definitions/JSONSchema7Definition",
            },
            Object {
              "items": Object {
                "$ref": "#/definitions/JSONSchema7Definition",
              },
              "type": "array",
            },
          ],
        },
        "maxItems": Object {
          "type": "number",
        },
        "maxLength": Object {
          "type": "number",
        },
        "maxProperties": Object {
          "type": "number",
        },
        "maximum": Object {
          "type": "number",
        },
        "minItems": Object {
          "type": "number",
        },
        "minLength": Object {
          "type": "number",
        },
        "minProperties": Object {
          "type": "number",
        },
        "minimum": Object {
          "type": "number",
        },
        "multipleOf": Object {
          "type": "number",
        },
        "not": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "oneOf": Object {
          "items": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "array",
        },
        "pattern": Object {
          "type": "string",
        },
        "patternProperties": Object {
          "additionalProperties": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "object",
        },
        "properties": Object {
          "additionalProperties": Object {
            "$ref": "#/definitions/JSONSchema7Definition",
          },
          "type": "object",
        },
        "propertyNames": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "readOnly": Object {
          "type": "boolean",
        },
        "required": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "then": Object {
          "$ref": "#/definitions/JSONSchema7Definition",
        },
        "title": Object {
          "type": "string",
        },
        "type": Object {
          "anyOf": Array [
            Object {
              "$ref": "#/definitions/JSONSchema7TypeName",
            },
            Object {
              "items": Object {
                "$ref": "#/definitions/JSONSchema7TypeName",
              },
              "type": "array",
            },
          ],
        },
        "uniqueItems": Object {
          "type": "boolean",
        },
        "writeOnly": Object {
          "type": "boolean",
        },
      },
      "type": "object",
    },
    "JSONSchema7Array": Object {
      "additionalProperties": undefined,
      "properties": Object {
        "length": Object {
          "type": "number",
        },
      },
      "required": Array [
        "length",
      ],
      "type": "object",
    },
    "JSONSchema7Definition": Object {
      "anyOf": Array [
        Object {
          "$ref": "#/definitions/JSONSchema7",
        },
        Object {
          "type": "boolean",
        },
      ],
      "description": "JSON Schema v7",
    },
    "JSONSchema7Object": Object {
      "additionalProperties": Object {
        "$ref": "#/definitions/JSONSchema7Type",
      },
      "type": "object",
    },
    "JSONSchema7Type": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "type": "number",
        },
        Object {
          "type": "boolean",
        },
        Object {
          "$ref": "#/definitions/JSONSchema7Object",
        },
        Object {
          "$ref": "#/definitions/JSONSchema7Array",
        },
        Object {
          "type": "null",
        },
      ],
      "description": "Primitive type",
    },
    "JSONSchema7TypeName": Object {
      "description": "Primitive type",
      "enum": Array [
        "string",
        "number",
        "integer",
        "boolean",
        "object",
        "array",
        "null",
      ],
      "type": "string",
    },
    "JSONSchema7Version": Object {
      "description": "Meta schema

Recommended values:
- 'http://json-schema.org/schema#'
- 'http://json-schema.org/hyper-schema#'
- 'http://json-schema.org/draft-07/schema#'
- 'http://json-schema.org/draft-07/hyper-schema#'",
      "type": "string",
    },
  },
}
`;

exports[`Use schemaGenerator to parse OmitExample.ts 3`] = `"ISchema"`;

exports[`Validating with ajv for ComplexExample.ts 1`] = `"x should be equal to one of the allowed values"`;

exports[`Validating with ajv for ComplexExample.ts 2`] = `"x.query.id should be number, x should have required property 'body', x.params.e should be equal to one of the allowed values"`;

exports[`Validating with ajv for Example.ts 1`] = `"x.email should match format \\"email\\""`;

exports[`Validating with ajv for Example.ts 2`] = `"x.answer should be number"`;

exports[`Validating with ajv for Example.ts 3`] = `"x.email should match format \\"email\\""`;

exports[`Validating with ajv for Example.ts 4`] = `"x.answer should be number"`;

exports[`Validating with ajv for OmitExample.ts 1`] = `"x.schema should be object"`;
